icu4c/source/common/uresbund.cpp

```diff
@@ -3145,10 +3174,33 @@ ures_getFunctionalEquivalent(char *result, int32_t resultCapacity,
             }
         }
         
-        subStatus = U_ZERO_ERROR;
-        
-        uprv_strcpy(found, parent);
-        uloc_getParent(found,parent,1023,&subStatus);
+        UBool haveFound = false;
+        // At least for collations which may be aliased, we need to use the VALID locale
+        // as the parent instead of just truncating, as long as the VALID locale is not
+        // root and has a different language than the parent. Use of the VALID locale
+        // here is similar to the procedure used at the end of the previous do-while loop
+        // for all resource types.
+        if (res != NULL && uprv_strcmp(resName, "collations") == 0) {
+            subStatus = U_ZERO_ERROR;
+            const char *validLoc = ures_getLocaleByType(res, ULOC_VALID_LOCALE, &subStatus);
+            if (U_SUCCESS(subStatus) && validLoc != NULL && validLoc[0] != 0 && uprv_strcmp(validLoc, "root") != 0) {
```

Comment:
The check for validLoc != "root" may be unnecessary in open-source ICU (not entirely sure), but it certainly does not hurt anything. Apple ICU retains use of "root" as a name where open-source ICU may use empty string...

Reply:
Thanks; if it doesn't hurt, I'm going to leave it.